### PR TITLE
[x86/Linux] WIP, Enable FEATURE_FIXED_OUT_ARGS

### PR DESCRIPTION
**Pull request by seanshpark**
This is initial patch to enable FEATURE_FIXED_OUT_ARGS for x86/Linux

Enable `UNIX_X86_ABI_FOA` line in src/inc/corabi.h to test.

### CODE REVIEW

**Comment by seanshpark** [comment]
Empty Main program works but caught on assert on lower.cpp:975

**Comment by seanshpark** [comment]
CC @YongseopKim

**Comment by seanshpark** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test please
@dotnet-bot test Windows_NT x86 Checked Build and Test please

**Comment by seanshpark** [comment]
@dotnet-bot test Windows_NT x86 Checked Build and Test please

**Comment by seanshpark** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test please
@dotnet-bot test OSX x64 Checked Build and Test please

**Comment by seanshpark** [comment]
Had to manually squash to resolve conflicts with master.

**Comment by seanshpark** [comment]
@dotnet-bot test Windows_NT x86 Checked Build and Test please

**Comment by seanshpark** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test please
@dotnet-bot test Ubuntu x64 Checked Build and Test please

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -2442,6 +2442,71 @@ void fgArgInfo::EvalArgsToTemps()
...
+
+    for (curInx = 0; curInx < argCount; curInx++)
+    {
+        fgArgTabEntryPtr curArgTabEntry = argTable[curInx];
+        if (curArgTabEntry->numSlots > 0)
+        {
+            // The argument may be REG_STK or constant or register that goes to stack
+            assert(nextSlotNum >= curArgTabEntry->slotNum);
+
+            numSlots += curArgTabEntry->numSlots;
```
Comment (wateret) [inline src/jit/morph.cpp]: Looks like this local variable `numSlots` is unused.

Comment (seanshpark) [inline src/jit/morph.cpp]: Thanks, it's from copy of previous stack alignment. Need to remove this and some comments also.

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build

**Comment by seanshpark** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by seanshpark** [comment]
@jkotas , @BruceForstall , could you please take a look?
- Feature is disabled by default. `UNIX_X86_ABI_FOA` is added to new `coreabi.h` as a switch, cause this seems to make at least files to recompile when switch enable/disable feature for comparison. If you would prefer to move to cmake or other place, please let me know.
- Although I don't think it's finished but I've done what I can do for now. Please guide me for incorrect or strange changes.
- `TODO-FOA` commented codes are codes that doesn't execute with unit test. One in `codegenxarch.cpp` runs with `UNIX_X86_ABI_FOA` disabled but doesn't when enabled. Condition (or state) doesn't seem to be exact. The other one in `codegencommon.cpp` I couldn't figure out how to run this. If you have any test codes, can you share them with me?
- There are two major things that are still under development besides this PR, hoisting nested call (in separate PR when ready, regardless of FIXED_OUT_ARGS) and fixing GCInfo (#11014).

**Comment by seanshpark** [comment]
Rebased to resolve conflicts. Addition to previous comments;
- Feature is disabled by default. You can turn it on in target.h (https://github.com/dotnet/coreclr/pull/10034/commits/ac8a383389dbd13c097402969574b6149ca6af2f#diff-c38dcfe374e920d3eba6decb15f5e6e1R362) by setting `FEATURE_FIXED_OUT_ARGS` to 1
- Previous `UNIX_X86_ABI_FOA` was removed as it's no longer needed.

**Comment by seanshpark** [comment]
@jkotas , @BruceForstall , it's been a while so, could you please take a look?

**Comment by BruceForstall** [comment]
@seanshpark ok, I'll look soon.

**Comment by BruceForstall** [comment]
@seanshpark I apologize I haven't been able to get to this, and I'm going on vacation for a week. Perhaps @RussKeldorph can get someone to look, maybe @pgavlin .

**Comment by pgavlin** [comment]
I will take a look at this this evening or tomorrow.

**Comment by seanshpark** [comment]
@BruceForstall , Thank you and have a wonderful vacation!

**Comment by pgavlin** [comment]
I'm not certain that I have enough context here for a proper review. Is there an explainatino somewhere re: why we're enabling `FEATURE_FIXED_OUT_ARGS` for x86/Linux?

**Comment by seanshpark** [comment]
@pgavlin , Please refer #10012. Please take your time.

**Comment by karelz** [comment]
What is status of this PR? Are we blocked on code review?
There hasn't been any activity for 3 months :(

**Comment by BruceForstall** [comment]
@karelz Yes, blocked on code review. I'd asked (elsewhere; can't find it now) and this apparently isn't blocking their use of Linux/x86.

**Comment by karelz** [comment]
If there is not near-term plan to finish the PR, I would suggest to close it and reopen when we are ready ...

**Comment by karelz** [comment]
Closing for now, feel free to reopen when you're ready to work on it again ...

**Pull request closed without merging**